src/algebra/module/zlattice.lean

```diff
@@ -0,0 +1,234 @@
+* For `b` a basis of `E`, then `L : submodule.span ℤ (set.range b)` is a ℤ-lattice of `E`.
+* As `L : add_subgroup E` with the additional properties:
+  `(hd : ∀ r : ℝ, (L ∩ (metric.closed_ball 0 r)).finite)`, that is `L` is discrete
+  `(hs : submodule.span ℝ (L : set E) = ⊤)`, that is `L` spans `E` over `K`.
+
+## Main result
+* `zspan.is_add_fundamental_domain`: for `L : submodule.span ℤ (set.range b)` a ℤ-lattice, proves
+ that the set defined by `zspan.fundamental_domain` is indeed a fundamental domain.
+
+-/
+
+open_locale classical
+
+noncomputable theory
+
+section zspan
+
+open measure_theory measurable_set submodule
+
+variables {E ι : Type*}
+
+section basis
+
+variables {R K : Type*} [comm_ring R] [division_ring K] [algebra R K] [add_comm_group E]
+  [module K E] [module R E] [is_scalar_tower R K E] [no_zero_smul_divisors R K]
+variables (b : basis ι K E)
+
+variable (R)
+
+/-- The R-lattice spanned by `b` admits `b` as a R-basis. -/
+def zspan.basis : basis ι R (span R (set.range b)) :=
+basis.span (b.linear_independent.restrict_scalars (smul_left_injective R (ne_zero.ne 1)))
+
+@[simp]
+lemma zspan.basis_apply (i : ι) : (zspan.basis R b i : E) = b i :=
+  by simp only [zspan.basis, basis.span_apply]
+
+@[simp]
+lemma zspan.repr_apply (m : span R (set.range b)) [finite ι] (i : ι)  :
+  algebra_map R K ((zspan.basis R b).repr m i) = b.repr m i :=
+begin
+  casesI nonempty_fintype ι,
+  rw ← congr_arg (coe : _ → E) (basis.sum_repr (zspan.basis R b) m),
+  simp_rw [coe_sum, coe_smul_of_tower, zspan.basis_apply, linear_equiv.map_sum,
+    ← is_scalar_tower.algebra_map_smul K, b.repr.map_smul, basis.repr_self, algebra_map_smul,
+    finsupp.smul_single, finset.sum_apply', algebra.algebra_map_eq_smul_one, finsupp.single_apply,
+    finset.sum_ite_eq', finset.mem_univ],
+  refl,
+end
+
+lemma zspan.mem_span_iff [finite ι] (m : E) :
+  m ∈ span R (set.range b) ↔ ∀ i, ∃ c : R, b.repr m i = algebra_map R K c :=
+begin
+  casesI nonempty_fintype ι,
+  split,
+  { exact λ hm i, ⟨(zspan.basis R b).repr ⟨m, hm⟩ i, (zspan.repr_apply R b ⟨m, hm⟩ i).symm⟩, },
+  { intros h,
+    rw ← b.sum_repr m,
+    refine sum_mem (λ i _ , _),
+    rw [(h i).some_spec, @is_scalar_tower.algebra_map_smul R K],
+    exact smul_mem _ _ (subset_span (set.mem_range_self i)), }
+end
+
+end basis
+
+section normed_lattice_field
+
+variables {K : Type*} [normed_lattice_field K]
+variables [normed_add_comm_group E] [normed_space K E]
+variables (b : basis ι K E)
+
+/-- The fundamental domain of the ℤ-lattice spanned by `b`. See `zspan.is_add_fundamental_domain`
+for the proof that it is the fundamental domain. -/
+def zspan.fundamental_domain : set E := { m | ∀ i : ι, b.repr m i ∈ set.Ico (0 : K) 1 }
+
+variables [floor_ring K]
+
+section fintype
+
+variable [fintype ι]
+
+/-- The map that sends a vector of `E` to the element of the ℤ-lattice spanned by `b` obtained
+by rounding down its coordinates on the basis `b`. -/
+def zspan.floor_map : E → span ℤ (set.range b) :=
```

Comment:
Can you show this agrees with `int.floor` when considering `ℤ` as a lattice inside a `floor_ring`?

Reply:
I would say you're missing API then 😁

Reply:
I am not sure I see the kind of lemmas that you have in mind...

Reply:
I want you to prove, for the correct basis `b : basis unit K K`
```
@[simp, norm_cast] lemma coe_floor_self (m : K) : ↑(zspan.floor k) = ⌊k⌋
```

Reply:
This is what I proved: 
```lean
section unique

variable [unique ι]

@[simp] lemma coe_floor_self (k : K) : (floor (basis.singleton ι K) k : K) = ⌊k⌋ :=
(basis.ext_elem_iff _).mpr
  (λ _, by { rw [floor_single, basis.singleton_repr, basis.singleton_repr], })

@[simp] lemma coe_fract_self (k : K) : (fract (basis.singleton ι K) k : K) = int.fract k :=
(basis.ext_elem_iff _).mpr
  (λ _, by { rw [fract_single, basis.singleton_repr, basis.singleton_repr], })

end unique
```
Note that Lean complains if I add the `norm_cast` tag: 
```
norm_cast: badly shaped lemma, rhs can't start with coe
```